site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+
+	// BackendRefs defines the backend(s) where matching requests should be
+	// sent.
+
+	// If unspecified or invalid (refers to a non-existent resource or a Service
+	// with no endpoints), the rule performs no forwarding. If there are also no
+	// filters specified that would result in a response being sent, a gRPC `UNAVAILABLE`
+	// status is returned. `UNAVAILABLE` responses must be sent so that the overall
+	// weight is respected; if an invalid backend is requested to have 80% of
+	// requests, then 80% of requests must get a `UNAVAILABLE` instead.
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []GRPCBackendRef `json:"backendRefs,omitempty"`
+}
+
+// GRPCRouteMatch defines the predicate used to match requests to a given
+// action. Multiple match types are ANDed together, i.e. the match will
+// evaluate to true only if all conditions are satisfied.
+//
+// For example, the match below will match a gRPC request only if its service
+// is `foo` AND it contains the `version: v1` header:
+//
+// ```
+// match:
+//   method:
+//     type: Exact
+//     service: "foo"
+//   headers:
+//   - name: "version"
+//     value "v1"
+// ```
+type GRPCRouteMatch struct {
+	// Path specifies a gRPC request service/method matcher. If this field is not
+	// specified, all services and methods will match.
+	//
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"path,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCPathMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method..
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"value,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Method *string `json:"value,omitempty"`
+}
+
+// MethodMatchType specifies the semantics of how gRPC methods and services should be compared.
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
```

Comment:
Conformance levels?

Reply:
This is specified a bit higher up, in the `GRPCMethodMatch` struct.

Reply:
Do we need match types here? If so, do they apply to both Service and Method? Should we split those into separate matches?

Reply:
> Do we need match types here?

I think so. See my use cases for regex matches [here](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).

> If so, do they apply to both Service and Method?

Yes.

> Should we split those into separate matches?

I don't see the value add in doing this. Again, see [this comment](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).